### PR TITLE
build the menu, update url upon click

### PR DESCRIPTION
**Pull request by qlj-lijuan**
- build the gni menu using react js
- add the function of updating the url once the statvar is selected

### CODE REVIEW

File: static/css/gni.css [hunk 1]
```diff
@@ -291,7 +291,7 @@ body {
 }
```
Comment (shifucun) [inline static/css/gni.css:291]: Can you make a screenshot on how the current GNI would look like with the css change? Trying not to make too big a difference for it.

File: static/js/dev.ts [hunk 2]
```diff
@@ -44,6 +49,11 @@ window.onload = function () {
     chartElem.style.width = width + "px";
     return containerId;
   }
+  let width_menu = 600
```
Comment (shifucun) [inline static/js/dev.ts:52]: const widthMenu = 600;

Try to use const as much as possible; js style guide uses camelCase for variable and functions

File: static/js/pv_tree.tsx [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+
+interface NodePropType {
+  title: string; 
+  selected: string;
+  expanded: string;
+  count: number;
+  children: NodeType[];
+  type: string;
+  argString: string;
+  updateurl: (string, boolean) => void;
```
Comment (shifucun) [inline static/js/pv_tree.tsx:29]: updateUrl

File: static/js/pv_tree.tsx [hunk 4]
```diff
@@ -0,0 +1,163 @@
+import React, {Component} from "react";
+import hierarchy from "../../tools/pv_tree_generator/hierarchy_golden.json";
+//import hierarchy from "../data/hierarchy.json";
+
+// Fields of each node from json file
+interface NodeType {
+  show: string; 
+  title: string; 
+  selected: string;
```
Comment (shifucun) [inline static/js/pv_tree.tsx:9]: selected, expanded, placeTypes can be removed?

File: static/js/pv_tree.tsx [hunk 5]
```diff
@@ -0,0 +1,163 @@
+import React, {Component} from "react";
+import hierarchy from "../../tools/pv_tree_generator/hierarchy_golden.json";
+//import hierarchy from "../data/hierarchy.json";
+
+// Fields of each node from json file
+interface NodeType {
```
Comment (shifucun) [inline static/js/pv_tree.tsx:6]: Can call this NodeData?

File: static/js/pv_tree.tsx [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+          {child}
+        </li>
+      </ul>
+    )
+  }
+}
+
+interface MenuPropType{
+  search: boolean;
+  updateurl: (string, bool) => void;
```
Comment (shifucun) [inline static/js/pv_tree.tsx:116]: updateUrl

File: static/js/pv_tree.tsx [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+  type: string; // Property or Value
+  argString: string;// for statvars, argString = dcid
+}
+
+interface NodePropType {
+  title: string; 
+  selected: string;
+  expanded: string;
+  count: number;
+  children: NodeType[];
```
Comment (shifucun) [inline static/js/pv_tree.tsx:26]: Not quite sure how react handles the the memory of the Props, but the Menu object seems very heavy as each node contains the data for its subtree. Can you think if it is possible to create the tree without passing the data along the way?

Try https://reactjs.org/docs/react-api.html#createelement which takes the children elements as the third argument...

File: static/js/pv_tree.tsx [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+  render() {
+    let checkboxImg;
+    let expandImg;
+    let child; 
+    if (this.props.type === "value") {
+      checkboxImg = <button className={this.state.checked? 
+                                        "checkbox checked":"checkbox"}  
+                              onClick={this._handleCheckboxClick}/>};
+
+    if (this.state.expanded) {
```
Comment (shifucun) [inline static/js/pv_tree.tsx:67]: this is quite interesting, so you only create node when it is expanded, then the actual created nodes should be much less than before where we build the entire tree?

File: static/js/util.js [hunk 9]
```diff
@@ -493,6 +494,38 @@ function randDomId() {
...
 
+/**
+ * add or delete statvars from url
+ *
+ * @param {string} dcid of statvar
+ * @param {boolean} add = True, delete = False
+ * @return {string} The line chart URL.
+ */
+
+function updateUrlStatsVar(statvar, add){
```
Comment (shifucun) [inline static/js/util.js:505]: this would be a good case to add unit test

**Comment by shifucun** [comment]
Try run the test locally and fix lint error and the "Codacy" lint error.

File: static/js/pv_tree.tsx (hunk 7 above)
Comment (shifucun) [reply static/js/pv_tree.tsx:26]: Thinking a bit more I guess what you do is okay. Even all nodes are expanded, the total of all the props object is similar to the original object size. And if it's true that you only create the Node that is expanded, then the actual props passed down is very small. Creating node on the fly is definitely a good improvement!

File: static/css/gni.css (hunk 1 above)
Comment (qlj-lijuan) [reply static/css/gni.css:291]: Here's the screenshot.
https://drive.google.com/file/d/18AKyoYPRmswHM4cU3jRgu9Wyv4bPmAT6/view?usp=sharing

File: static/js/dev.ts (hunk 2 above)
Comment (qlj-lijuan) [reply static/js/dev.ts:52]: Got it. Thanks.

File: static/js/pv_tree.tsx (hunk 3 above)
Comment (qlj-lijuan) [reply static/js/pv_tree.tsx:29]: done.

File: static/js/pv_tree.tsx (hunk 4 above)
Comment (qlj-lijuan) [reply static/js/pv_tree.tsx:9]: We actually don't need this interface. Removed it.

File: static/js/pv_tree.tsx (hunk 5 above)
Comment (qlj-lijuan) [reply static/js/pv_tree.tsx:6]: removed it.

File: static/js/util.js (hunk 9 above)
Comment (qlj-lijuan) [reply static/js/util.js:505]: done.

File: static/js/pv_tree.tsx (hunk 8 above)
Comment (qlj-lijuan) [reply static/js/pv_tree.tsx:57]: right.

File: static/js/pv_tree.tsx [hunk 10]
```diff
@@ -0,0 +1,169 @@
+import React, { Component } from "react";
+import hierarchy from "../../tools/pv_tree_generator/hierarchy_golden.json";
+//import hierarchy from "../data/hierarchy.json";
+
+interface NodePropType {
+  title: string;
+  selected: string;
+  expanded: string;
```
Comment (shifucun) [inline static/js/pv_tree.tsx:8]: do think selected and expanded can be removed? It's okay to do in another PR if so, and also updated the hierachy.json

File: static/js/pv_tree.tsx [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+    this._handleExpandClick = this._handleExpandClick.bind(this);
+    this.state = {
+      checked: props.selected === "yes" ? true : false,
+      expanded: props.selected === "yes" ? true : false,
+      nodePath: props.nodePath + "," + props.title,
+    };
+  }
+
+  render() {
+    let checkboxImg;
```
Comment (shifucun) [inline static/js/pv_tree.tsx:36]: good to have type like: let checkboxImg: HTMLInputElement for all the let varibles

File: static/js/util.test.js [hunk 12]
```diff
@@ -0,0 +1,16 @@
+import { updateUrlStatsVar } from "./util.js";
+
+test("update Url statsvar", () => {
```
Comment (shifucun) [inline static/js/util.test.js:3]: nice!

File: static/js/dev.ts [hunk 13]
```diff
@@ -82,7 +92,8 @@ window.onload = function () {
       new DataPoint("2013", 2000),
     ]),
   ];
-
+  
+ //ReactDOM.render(React.createElement(page),document.getElementById(containerId));
```
[the remark this answers is not in the document]
Comment (qlj-lijuan) [reply static/js/dev.ts:96]: done.

File: static/js/pv_tree.tsx (hunk 10 above)
Comment (qlj-lijuan) [reply static/js/pv_tree.tsx:8]: I'll do it in another PR and update the hierarchy.json.

File: static/js/pv_tree.tsx (hunk 11 above)
Comment (qlj-lijuan) [reply static/js/pv_tree.tsx]: done.

File: static/css/gni.css [hunk 14]
```diff
@@ -322,10 +322,10 @@ body {
   font-size: 0.95rem;
   margin: 0 -24px;
   margin-left: 6px;
-  padding: 3px 24px 3px 6px;
+  padding: 0 24px 0 6px;
   border-top-right-radius: 6px;
   border-bottom-right-radius: 6px;
-  border-left: 4px solid #ccc;
+  /* border-left: 4px solid #ccc;  */
```
Comment (beets) [inline static/css/gni.css:328]: Was this accidentally removed?

File: static/js/pv_tree.tsx [hunk 15]
```diff
@@ -0,0 +1,169 @@
+import React, { Component } from "react";
+import hierarchy from "../../tools/pv_tree_generator/hierarchy_golden.json";
+//import hierarchy from "../data/hierarchy.json";
```
Comment (beets) [inline static/js/pv_tree.tsx:3]: Please delete if not needed anymore

File: static/js/pv_tree.tsx [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+  nodePath: string;
+}
+
+class Node extends Component<NodePropType, NodeStateType> {
+  constructor(props: NodePropType) {
+    super(props);
+    this._handleCheckboxClick = this._handleCheckboxClick.bind(this);
+    this._handleExpandClick = this._handleExpandClick.bind(this);
+    this.state = {
+      checked: props.selected === "yes" ? true : false,
```
Comment (beets) [inline static/js/pv_tree.tsx:29]: This is minor, but it should work to simply do:

checked: props.selected === "yes"

File: static/js/pv_tree.tsx [hunk 17]
```diff
@@ -0,0 +1,169 @@
...
+        />
+      );
+    }
+
+    if (this.state.expanded) {
+      expandImg = (
+        <img
+          className="right-caret transform-down"
+          src="../images/right-caret.png"
+          width="12px"
```
Comment (beets) [inline static/js/pv_tree.tsx:53]: Could you move this parameter to the stylesheet? Looks like if we apply it to the right-caret class, it wouldn't be needed below on line 80 as well.

File: static/js/util.js [hunk 18]
```diff
@@ -493,6 +494,37 @@ function randDomId() {
...
 }
 
+/**
+ * add or delete statvars from url
+ *
+ * @param {string} dcid of statvar
+ * @param {boolean} add = True, delete = False
+ * @return {string} The line chart URL.
+ */
+
```
Comment (beets) [inline static/js/util.js:504]: Since this comment applies to the function, the blank line isn't necessary.

File: static/js/util.js [hunk 19]
```diff
@@ -493,6 +494,37 @@ function randDomId() {
...
+function updateUrlStatsVar(statvar, add) {
```
Comment (beets) [inline static/js/util.js:505]: Minor, but if you rename add to should_add or should_be_added or is_statvar_added, then it's very clear it's a boolean, and line 511 can be very clearly read as: if (is_statvar_added) { ... }

File: static/css/gni.css (hunk 14 above)
Comment (qlj-lijuan) [reply static/css/gni.css:328]: This is removed deliberately, because otherwise there're two solid lines on the left for each level. I'm no sure if there's a better way to get rid of those lines.

File: static/js/pv_tree.tsx (hunk 15 above)
Comment (qlj-lijuan) [reply static/js/pv_tree.tsx]: done.

File: static/js/pv_tree.tsx (hunk 16 above)
Comment (qlj-lijuan) [reply static/js/pv_tree.tsx]: done.

File: static/js/pv_tree.tsx (hunk 17 above)
Comment (qlj-lijuan) [reply static/js/pv_tree.tsx]: done.

File: static/js/util.js (hunk 18 above)
Comment (qlj-lijuan) [reply static/js/util.js]: got it.

File: static/js/util.js (hunk 19 above)
Comment (qlj-lijuan) [reply static/js/util.js]: done.
